File: Olympiad/POI/official/03-X/kry/zawody/pas/krysztal.pas
unit krysztal; { krysztal.pas }

{* interefejs komunikacji z aparatur± 
 * w zadaniu Kryszta³
 * w jêzyku Pascal
 *
 * obja¶nienia w tre¶ci zadania
 *}

interface

function rozmiar: integer;

function zamroz(x, y : integer) : integer;

procedure niestabilna (x1, y1, x2, y2 : integer);

implementation

{* uproszczony modu³ sterujacy aparatura
 * do testowania rozwiazan przez zawodnikow
 * wczytuje dane o krysztale ze standardowego wejscia
 *}

const MAX_PROG = 1000; { Uwaga: W zadaniu ograniczenie wynosi MAX_N. }
const MAX_N = 10000; 

var 
  kry: array[1..MAX_PROG, 1..MAX_PROG] of integer;
  n : integer;

procedure wczytaj; 
var pom, i, j: integer;
begin
  read(n);
	if (n<3) or (n>MAX_N) then begin
	  writeln('Niepoprawna wartosc n.');
		halt(1);
	end;
	if (n > MAX_PROG) then begin
	  writeln('Poprawna wartosc n, ale za duza dla tego programu.');
		halt(1);
	end;
	for i:=1 to n do 
	  for j:=1 to n do begin
		  read(pom);
			if (pom=0) or (abs(pom)>2) then begin
			  writeln('Wczytano niepoprawna wartosc (', pom, ') stanu atomu (', i, ',', j, ').');
				halt(1);
			end;
			kry[i, j] := pom;
		end;
end;

var ile : integer;

function zamroz(x, y : integer) : integer;
begin
  if (x<1) or (x>n) or (y<1) or (y>n) then begin
	  writeln('Wspolrzedne atomu spoza dopuszczalnego zakresu');
		halt(1);
	end;
  if (x>1) and (x<n) and (y>1) and (y<n) then inc(ile);
  zamroz := kry[x][y];
end;

procedure niestabilna(x1, y1, x2, y2 : integer);
begin
  if (x1<1) or (x1>n) or (y1<1) or (y1>n) or 
	   (x2<1) or (x2>n) or (y2<1) or (y2>n) then begin
	  writeln('Wspolrzedne atomu spoza dopuszczalnego zakresu');
		halt(1);
	end;
	
  writeln('Program dokonal ', ile, ' zamrozen (limit=',3*n,').');
  if (abs(x2-x1) <= 1) and (abs(y2-y1) <= 1) and (kry[x1,y1] = -kry[x2, y2]) then begin
    writeln('Poprawna odpowiedz.');	
		halt(0);
	end else begin
    writeln('Niepoprawna odpowiedz.');	
	  halt(1);
	end;
end;

var initialized : boolean;

function rozmiar: integer; 
begin
  if (not initialized) then begin
    wczytaj;
    initialized := true;
    ile := 0;
  end else begin
    writeln('Funkcja rozmiar moze byc wywolana najwyzej raz'); 
  end;
  rozmiar := n;
end;

begin 
  initialized := false;
end.

File: Olympiad/POI/official/2003/kry/zawody/c/krysztal.c
/* Program sprawdzaj±cy dla zawodników.
// Pozwala na korzystanie z nieco mniejszej maksymalnej warto¶ci N
// Nale¿y skompilowaæ w³asny plik z tym modu³em (krysztal.h / krysztal.cpp)
// Pobiera dane ze standrdowego wej¶cia
// U¿ycie: ./kry < nazwa_pliku
// Format pliku:
// n
// [1;1] [2;1] ... [n;1]
// [1;2] [2;2] ... [n;2]
// ...
// [1;n] [2;n] ... [n;n]
// gdzie [x;y] oznacza stan kryszta³u o wspó³rzêdnych (x,y)
*/
#include <stdio.h>
#include <stdlib.h>

#include "krysztal.h"

#define MAX_PROG 1000 
/* Uwaga: W zadaniu ograniczenie wynosi MAX_N. */
#define MAX_N 10000 

static int kry[MAX_PROG+1][MAX_PROG+1]; /* indeksujemy od 1 do n */
static int n;

static int wczytajLiczbe() {
  int pom;
  if (scanf("%d", &pom) != 1) {
    fprintf(stderr, "Blad odczytu wejscia!\n");
    exit(1);
  }
  return pom;
}

static void wczytaj() {
  int pom, i, j;
  n = wczytajLiczbe();
  if (n < 3 || n > MAX_N) {
    fprintf(stderr, "Niepoprawna wartosc n.\n");
    exit(1);
  }
  if (n > MAX_PROG) {
    fprintf(stderr, "Poprawna wartosc n, ale za duza dla tego programu.\n");
    exit(1);
  }
  for (i = 1; i <= n; ++i)
    for (j = 1; j <= n; ++j) {
      pom = kry[j][i] = wczytajLiczbe();
      if (pom != -1 && pom != 1 && pom != -2 && pom != 2) {
        fprintf(stderr, "Wczytano niepoprawna wartosc (%d) stanu atomu (%d, %d).\n", pom, j, i);
        exit(1);
      }
    }
}

static int ile;

int zamroz(int x, int y) {
  if (x<1 || x>n || y<1 || y>n) {
    fprintf(stderr, "Wspolrzedne atomu poza zakresem");
    exit(1);
  }
  if (x>1 && x<n && y>1 && y<n) ile++;
  return kry[x][y];
}

void niestabilna(int x1, int y1, int x2, int y2) {
  if (x1<1 || x1>n || y1<1 || y1>n || x2<1 || x2>n || y2<1 || y2>n) {
    fprintf(stderr, "Wspolrzedne atomu poza zakresem");
    exit(1);
  }

  fprintf(stderr, "Program dokonal %d zamrozen (limit=%d).\n", ile, 3*n);

  if (abs(x2-x1) <= 1 && abs(y2-y1) <= 1 && kry[x1][y1] == -kry[x2][y2]) {
    fprintf(stdout, "Poprawna odpowiedz.\n");
    exit(0);
  } else {
    fprintf(stdout, "Niepoprawna odpowiedz.\n");
    exit(1);
  }
}

static int initialized = 0;

int rozmiar() {
  if (!initialized) {
    wczytaj(); /* z stdin */
    initialized = 1;
    ile = 0;
  } else {
    fprintf(stderr, "Funkcja rozmiar() mo¿e byæ wywolana najwyzej raz"); 
  }
  return n;
}

